Remove unused splitByTrace from groupbytraceprocessor, add tests to batchpersignal

LouisStAmour opened the pull request:
I was looking at groupbytraceprocessor to try and build a log processor that would collect an ordered ring buffer (circular queue) of logs using insertion sort and emit a sorted part of the buffer after some delay to fix #3090. This PR doesn't do that - while investigating the rather complicated (to me) control flow of groupbytraceprocessor, I discovered that the splitByTrace function it had was only referenced in tests. So this PR removes that function. In doing so, I ported existing tests from groupbytraceprocessor to batchpersignal and discovered that how Nil was handled changed when the batchpersignal function was added to groupbytraceprocessor.

As noted in the commit text: Note that previously groupbytraceprocessor expected Nil trace IDs to be removed, now they are kept. This might have further consequences for groupbytraceprocessor, I couldn’t find any tests for it that included Nil trace IDs. That said, the only consequence I can think of is that it would attempt to group all Nil trace IDs together, which is possibly a bug. I’d imagine the better behaviour would be to emit/pass through any traces with Nil trace IDs immediately? But the switch to using batchpersignal has already occurred, so fixing this is out of scope for this commit.

bogdandrutu commented on processor/groupbytraceprocessor/processor.go:
```diff
@@ -245,58 +245,4 @@ func (sp *groupByTraceProcessor) onTraceRemoved(traceID pdata.TraceID) error {
-
-	return result
-}
+}
```
New line.

bogdandrutu commented on pkg/batchpersignal/batchpersignal_test.go:
```diff
@@ -95,3 +157,65 @@ func TestSplitDifferentLogsIntoDifferentBatches(t *testing.T) {
+	assert.Equal(t, pdata.NewTraceID([16]byte{1, 2, 3, 4}), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).TraceID())
+	assert.Equal(t, secondLibrary.Name(), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).InstrumentationLibrary().Name())
+	assert.Equal(t, thirdLog.Name(), batches[1].ResourceLogs().At(0).InstrumentationLibraryLogs().At(0).Logs().At(0).Name())
+}
```
New line.

bogdandrutu reviewed (commented):
@jpkrohling can you also take a look?

bogdandrutu commented:
Please also fix lint issues:
```
make[1]: Entering directory '/home/circleci/project/pkg/batchpertrace'
golangci-lint run --allow-parallel-runners
batchpersignal_test.go:221: File is not `gofmt`-ed with `-s` (gofmt)
}
make[1]: *** [../../Makefile.Common:93: lint] Error 1
```

jpkrohling commented:
>while investigating the rather complicated (to me) control flow of groupbytraceprocessor

Indeed, it's now more complicated than it deserves. The original idea was to simplify the concurrency access to the underlying data structures, but I would probably do it differently today. 

>I discovered that the splitByTrace function it had was only referenced in tests

Good catch, this was probably a leftover of the change that moved this logic to the batch per signal library.

>discovered that how Nil was handled changed when the batchpersignal function was added to groupbytraceprocessor.

Batch per signal was introduced in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2549, to make the load balancing exporter work with logs as well. I believe this change might have been originated there. Either that or when the traceID changed from being a byte array to its own structure.

tigrannajaryan commented:
The linting failed and this seems stale now. Is there an intent to continue working on this PR?

LouisStAmour commented:
Happy to fix the PR, I've added it to my todo list for tomorrow. Sorry, I'm a bit new to more formal open source contributions. :)

jpkrohling commented:
This needs a rebase now.

LouisStAmour committed:
Remove unused splitByTrace from groupbytraceprocessor, add tests to batchpersignal

Note that previously groupbytraceprocessor expected Nil trace IDs to be removed, now they are kept. This might have further consequences for groupbytraceprocessor, I couldn’t find any tests for it that included Nil trace IDs. That said, the only consequence I can think of is that it would attempt to group all Nil trace IDs together, which is possibly a bug. I’d imagine the better behaviour would be to emit/pass through any traces with Nil trace IDs immediately? But the switch to using batchpersignal has already occurred, so fixing this is out of scope for this commit.

LouisStAmour commented:
My first force-push was a rebase against origin/main, the second fixed a gofmt error in one file. I couldn't find any lint errors or gofmt errors outside of that. Also ran gotidy and didn't notice any issues against processor/groupbytraceprocessor.

bogdandrutu commented:
Please fix the tests

LouisStAmour commented:
I've rebased against main in the hopes that tests will fix themselves. They seemed to fail due to a recent commit for oauth and some windows test failures.
